### PR TITLE
(Partially) Implemented Ioctl

### PR DESCRIPTION
**Pull request by youngfalcon**

### CODE REVIEW

**Comment by kapkic** [comment]
I believe this needs a merge from develop to get getuid/gid calls as the getuid tests are failing.

File: sysdeps/nacl/ioctl.c [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    if (fd < 0) {
+      __set_errno (EBADF);
+      return -1;
+    }
+
+    int arg = 0;
+    void* arg_ptr = &arg;
+    int result;
+
+    if (request == FIONBIO) {
```
Comment (rennergade) [inline sysdeps/nacl/ioctl.c:18]: is FIONBIO included from the ioctl header?

Comment (youngfalcon) [reply sysdeps/nacl/ioctl.c:18]: Yes. It is included from` \sysdeps\unix\bsd\bsd4.4\bits\ioctls.h` which is inclued through `<sys/ioctl.h>`

Comment (rennergade) [reply sysdeps/nacl/ioctl.c:18]: hmm I believe its actually coming from the x86 header which pulls in some asm definitions. Either way its there. Cool!

**Comment by rennergade** [comment]
Approved.

**Review by kapkic — APPROVED** [review]

File: sysdeps/nacl/ioctl.c [hunk 2]
```diff
@@ -0,0 +1,37 @@
+#include <errno.h>
+#include <stdarg.h>
+#include <sys/ioctl.h>
+
+#include <irt_syscalls.h>
+
+int __ioctl (int fd, unsigned long request, ...) {
```
Comment (jesings) [inline sysdeps/nacl/ioctl.c:7]: I don't think the use of a variadic function is either necessary or helpful... I think just having a 3rd void* argument would be better? unless there is something I am missing.

**Review by jesings — COMMENTED** [review]
Mostly good! One comment.

File: sysdeps/nacl/ioctl.c (hunk 2 above)
Comment (youngfalcon) [reply sysdeps/nacl/ioctl.c:7]: I wanted to keep the implementation the same as the POSIX standard as the Linux man-pages specified the function to be called this way.

The function can also be called without any parameters so I am not sure if putting the third argument there is the best approach.

Comment (jesings) [reply sysdeps/nacl/ioctl.c:7]: I think you're probably correct, the signature doesn't matter *that* much considering this is a library function but overall I think you're right that using variadic arguments is less confusing than the alternative here (even though variadic arguments in C are a total mess).

**Review by jesings — APPROVED** [review]

File: sysdeps/nacl/ioctl.c [hunk 3]
```diff
@@ -11,10 +11,18 @@ int __ioctl (int fd, unsigned long request, ...) {
...
     int arg = 0;
-    void* arg_ptr = &arg;
+    void* arg_ptr = &arg; //Replaced later for requests using an arg
     int result;
 
+    /*Right now, we are only passing down pointers to NaCl. Any non-pointer value
+      for arg_ptr will cause a sysaddr translation error in NaCl.
+      As a result, the arg_ptr should be set to &va_arg (ap, void *)
+      instead of va_arg (ap, void *) for requests that directly accept 
+      an arg value and not a pointer.*/
```
Comment (jesings) [inline sysdeps/nacl/ioctl.c:24]: Also mention that this should be the case for struct pointers too.

Comment (youngfalcon) [reply sysdeps/nacl/ioctl.c:24]: How does this apply to struct pointers? Thought va_arg (ap, void *) would be valid for struct pointers?

Comment (jesings) [reply sysdeps/nacl/ioctl.c:24]: Because we can't unpack them in the union the same way because they'll mess up the size/representation

Comment (youngfalcon) [reply sysdeps/nacl/ioctl.c:24]: Oh, yes you are right.

Comment (youngfalcon) [reply sysdeps/nacl/ioctl.c:24]: Changed the implementation at the safeposix side.

**Pull request merged**